Added a recipe status as per ldtteam/minecolonies-features#152

GaeaKat opened the pull request:
Closes ldtteam/minecolonies-features#152

# Changes proposed in this pull request:
- Adds an x of y line to the recipes screen detailing how many recipes used and what the max number is.

Review please

GaeaKat committed:
Added a recipe status as per ldtteam/minecolonies-features#152

GaeaKat committed:
Added a recipe status as per ldtteam/minecolonies-features#152

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=5918) <br/>All committers have signed the CLA.

GaeaKat commented:
![image](https://user-images.githubusercontent.com/690546/95921154-b064e080-0da8-11eb-9d4b-4b6a068348ce.png)
![image](https://user-images.githubusercontent.com/690546/95921167-b5c22b00-0da8-11eb-9908-3fbf39a1a02e.png)
![image](https://user-images.githubusercontent.com/690546/95921176-b955b200-0da8-11eb-9722-0e7e37dc20bb.png)

Mekle001 commented on src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingCrafter.java:
```diff
@@ -181,6 +181,16 @@ public static boolean canBuildingCanLearnMoreRecipes(final int buildingLevel, fi
+     * @param buildingLevel the building level
+     * @return maxinum number of recipes
+     */
+    public static int buildingMaxRecipes(final int buildingLevel)
```
If we're going to have a central place like this, we need to update the "canBuildingCanLearnMoreRecipes" code to use it, otherwise we have a chance of it getting out of sync.

Mekle001 replied:
thinking more about this, you probably need a static and an instance version of this code. I think the instance should be _public int getMaxRecipeCount()_

marchermans replied:
We should never ever have a static version of this code.
Please refactor this so that it is building/worker specific.

There is too much context required to make this calculation even remotely feasible as a static method.

Raycoms replied:
Yeah, ask the building in the view sync for the max and synch it over instead of this additional static method

Mekle001 commented on src/main/java/com/minecolonies/coremod/client/gui/WindowListRecipes.java:
```diff
@@ -147,6 +157,7 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
     public void onUpdate()
     {
         updateRecipes();
+        recipeStatus.setLabelText(LanguageHandler.format("com.minecolonies.coremod.gui.workerhuts.recipestatus",building.getRecipes().size(),AbstractBuildingCrafter.buildingMaxRecipes(building.getBuildingLevel())));
```
You should not use a static accessor here for buildingMaxRecipe. There is a possibility not every building has the same number of recipes, so you need it to be overridable.  A good example of this is the lumberjack, where the recipe count and max recipes are always the same.

MatthiasMann replied:
And it needs to take the research for more recipe slots into account.

marchermans replied:
Also constants!
All translationkeys go in TranslationKeys

marchermans reviewed (changes requested):
Remove the static method.
It is not context sensitive enough

Mekle001 committed:
Make request system use localized names in display (#5908)

someaddons committed:
Serveral fixes (#5928)

Adds Custom arrows for our entities, which do not scale in damage with movement and despawn quickly after beeing on the ground for a bit, avoiding fps lag from large battles with thousands of arrows.

Some raid balancing:
Raiders now have less armor and scale less in health
Colonies raidlevel grows more linearly than before and is a bit lower.
When less than 5% of raiders remain the raids now automatically end.

Fixes:
Training archers no longer deal damage when shooting
ToggleAI no longer toggles single blocks on doors and instead uses the doors own toggle functionality
Raiders no longer accidentally toggle doors
Fix guarding guards "Jiggling" because they didnt fix up their starting position for pathing
Fix pathing starting positions, now it checks for an existing collision box and if the citizen is actually within it so that e.g. standing next to a Trapdoor no longer assumes a pathing node above the trapdoor. Before it used entity positions which can vary too much, and caused some cases where citizens get stuck.
Fix players not always beeing added to the colony view subs after dim changes, they now check if they're added each time they enter a chunk claimed by a colony making it more resistant.

someaddons committed:
Delivery orders (#5921)

This adapts the way deliveries are assigned to deliveryman to a multi-factor comparison system.
The resolver for dmen now resides in a warehouse and resolves the request it gets to its delivermen in our case those of a single warehouse. So now we compare all assigned deliverymen of a warehouse to get the best fit for the request.
The resolver now ask the dmen for a score for adding a certain request and then compares the score to other dmen's scores. The score(lower is better) itself is combined of these factors:

Closeness: How close the target/source of the new request is compared to an existing one, in relation to its total travel length.
Priority(also contains age): Overtaking a request of a higher priority is worse.
Pickup/Delivery alternating: When alternating between pickups and deliveries with close targets the score gets a little bonus.
Index: The index we would add the new request at, higher the more request we overtake and the more request we have in total.
These result in lists on deliverymen of approx ~3 times closer deliveries, allowing the AI to also bunch up deliveries more nicer in the future.

Fixes visitors sometimes getting spawn on zero pos.
Fix raids not respecting distance to unloaded buildings
Fix shipsraids sometimes instantly despawning after killing spawners, now takes a day to do so.

someaddons committed:
Fix citizen sometimes failing to respawn (#5932)

Fix our respawn position to the Townhall when our spawnpos sometimes can't find a fitting place to spawn

HappyFacade committed:
Small improvements to readme (#5937)

Small improvements to readme

Mekle001 committed:
Fix furnace recipe loading (#5942)

Fix furnace recipe loading
Recently many furnace recipes quit resolving during teaching. It turns out that the recipe book has changed and it's now only returning a single recipe with a list of possible ingredients for things like charcoal and glass. This change now expands that out into recipes for each possible ingredient during load for lookup purposes.

Mekle001 committed:
Fix #5919 (#5943)

Treat Lanterns as non-solid/decorations

someaddons committed:
Fix citizens sometimes toggling gates/doors badly (#5950)

Improve weights of deliveryman to start offloading work to free deliverymen earlier

Raycoms committed:
Success/failure sounds (#5945)

Co-authored-by: someaddons <38401808+someaddons@users.noreply.github.com>

Mekle001 committed:
Track recipe usage by token in the recipe  manager (#5941)

* Track recipe usage by token in the recipe  manager

* remove contains checks

Co-authored-by: someaddons <38401808+someaddons@users.noreply.github.com>

Mekle001 committed:
Misc. Crafter fixes (#5951)

* Make smeltercrafters fuel better during smelting

* Better exclude assistant ingredients from the cook

* Fix #5948, and walking at the concrete mixer

Co-authored-by: someaddons <38401808+someaddons@users.noreply.github.com>

Mekle001 committed:
Fix #5880 (#5957)

This PR tweaks inventory handling in addition to removing the old one-stack crafting limitation that was causing #5880 to happen.

Mekle001 committed:
Multi-Output recipes, and reliable secondary outputs (#5905)

Add support for multi-output recipes, so that one recipe can create all variants of one timber block
Add support for custom recipes to require that the recipe 'exists' to be valid
This enables a multi-output recipe to replace a classic recipe in which the inputs match
When a multi-output replaces, it also looks for classic recipes who have outputs that are alternate in the multi-recipe and removes them
Add support for multi-recipes to cycle through the outputs in the recipe list in buildings
Add ability to capture actual secondary outputs (stay in grid) from recipes as they are being taught, This increases compatibility with many mod recipes, including aquaculture fish fillets.
Some minor tweaks to display of the request tree in buildings, to make crafting jobs a bit clearer
Fix Combat Academy and Archery to mark beds as unoccupied on wake up, letting the residents sleep the next night
use Localizable strings for Combat Academy and Archery
Ensure that multi recipes replace simple recipes, and that they update when they change
Change all of the lumberjack recipes to be multi-recipes. They are now Log -> [Stripped Log, Wood, Stripped Wood]

Mekle001 committed:
Fix inability to find fuel in building (#5960)

Fix the smelter crafters being unable to find fuel in the building (because they were looking in a furnace!)

Raycoms committed:
Fix/crafter prerequest (#5961)

This PR makes sure that the crafter reserves some stacks for pending requests such that the quantities for the followup requests may get delivered already.

Mekle001 committed:
Make special requests scale to processors (#5964)

This makes fuel requests for all furnace users scale 1 stack per furnace in the building.
This makes smeltables scale similarly.
Finally, this makes compostable requests scale 1 stack per composting barrel in the building

Raycoms committed:
fix 5962 and 5965 (#5966)

Fix Rs issues and log spam

GaeaKat committed:
Removed old recipe cap code

GaeaKat committed:
Merged Changes

GaeaKat committed:
Removed re-added old static method
